typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,1005 @@
+```
+Type `Cons` will add a value to the vector and increase its type-level length.
+
+```typescript
+type Cons<A, N extends Nat> = {a: A, v: Vec<A, N>};
+```
+
+And the main type `Vec` combines them with conditional type.
+
+Using `infer`, we are able to infer new type variable `R` from `N`.
+So, when `N` is a `Succ`, we can infer nested `R` from it and return `Cons<A, R>`.
+Otherwise, it is simply `Nil`.
+
+```typescript
+type Vec<A, N extends Nat> = N extends Succ<infer R> ? Cons<A, R> : Nil;
+```
+
+And here are examples of `Vec`'s usage.
+We define two helper function which build `Vec` values.
+And, as you can see, we can use it safely.
+
+```typescript
+const emptyVec: <A>() => Vec<A, Zero> = () => "nil";
+
+const pushVec: <A, N extends Nat>(a: A, v: Vec<A, N>) => Vec<A, Succ<N>> = (a, v) => {
+  return {a, v}
+};
+
+let empty: Vec<number, Zero> = emptyVec(); // Ok
+let oneElem: Vec<number, Succ<Zero>> = pushVec(1, empty); // Ok
+let twoElems: Vec<number, Succ<Succ<Zero>>> = pushVec(2, oneElem); // Ok
+let twoElemsInvalid: Vec<number, Succ<Zero>> = pushVec(2, oneElem); // error
+oneElem = twoElems; // error
+```
+
+This was a simple example what we can do with type level programming.
+Using such primitives, we can implement different operations on types and even a Fibonacci [implementation](https://mjj.io/2021/03/29/type-level-programming-in-typescript/), for example.
+Follow Mathias Jean Johansen's [blog post](https://mjj.io/2021/03/29/type-level-programming-in-typescript/) or above-mentioned Yuriy Bogomolov's [examples](https://github.com/YBogomolov/talk-typelevel-ts/blob/master/src/typelevel.d.ts).
+
+```typescript
+type Fibonacci<N, F0 = Zero, F1 = One> = {
+  acc: F0
+  n: N extends Succ<infer _> ? Fibonacci<Decrement<N>, F1, Add<F0, F1>> : never
+}[IfElse<Equals<Zero, N>, "acc", "n">]
+```
+
+### GADT and eDSL
+
+#### GADT
+
+Generalized algebraic data types in Haskell give us an ability to manually write types of the constructors.
+Having data type `D a` with type value `a` we are able to create constructors like `C :: Int -> D Int`.
+The key feature here is that Haskell may say us about equality of types `a` and `Int` here.
+
+The other great feature of this approach is that we can make nice eDSL using GADT.
+And again TypeScript will not give you such a good eDSL as a result, but it doesn't mean that we can't do it.
+
+TypeScript on the other hand doesn't have equality inference.
+Nevertheless, we can manually provide such equality to it as a value.
+
+We will use `fp-ts` here and later as it has a lot of build in useful types and functions.
+
+```typescript
+import { identity } from "fp-ts/lib/function";
+```
+
+The next step in `Equality` definition.
+To provide equality as a value we use Leibniz equality.
+Here is a simplified version of it.
+Interface `Equality` is a Hybrid Type which gives an ability for object act as a function.
+Then in `ArithExpr` we simplify `Equality<A, B>` into an identity `(a: A) => A`.
```

Comment:
I don't see Arith Expr

Reply:
Hm, how to write it here? To somehow explain `Equality` I need to say about things described below but describing such things without `Equality` explanation is not really good